### DISCUSSION TITLE
Jackson

### DISCUSSION THREAD

**Discussion by t1** [smallrye/smallrye-graphql#925] [General]
In #920, @phillip-kruger wrote:
> At some point we should start replacing JsonP with Jackson (in our implementation)

Why would we? All MicroProfile and JakartaEE runtimes must support JSON-P and JSON-B, only some support Jackson. I'd rather rely on a standard than on an implementation.

**Comment by phillip-kruger** [comment]
The biggest concern is performance. There is a general drive in Quarkus to do so.

**Reply by phillip-kruger** [reply]
We can of course make is pluggable, so the implementation can use either ?

**Reply by jmartisk** [reply]
If we make it pluggable, what will we do for cases when the public API exposes JSON-P/JSON-B classes? For example, https://github.com/eclipse/microprofile-graphql/blob/master/client/api/src/main/java/org/eclipse/microprofile/graphql/client/Response.java#L19 references `JsonObject`
I think creating an abstraction over that would be a PITA.

**Reply by phillip-kruger** [reply]
Yea, we will need to change that, maybe to a generic type, and the Response becomes an interface that the impl (Jackson/jsonp) will provide.

**Reply by t1** [reply]
JAXB is only an API (and Spec and TCK). The presumably slowness (do you have any numbers?) can only be in the implementation, yasson, can't it? So wouldn't any abstraction that we come up with actually be basically just a subset of JAXB?
I understand that Jackson works on becoming compatible with JAXB. We could provide an adapter until that is ready.

**Reply by mswiderski** [reply]
@t1 I think you mean `JSON-B` and not `JAXB`, right?

**Reply by t1** [reply]
Yes, of course... I always confuse those two. Thanks for clarifying!

**Comment by mswiderski** [comment]
I am also interested in this discussion if there will be any flavour that would allow to rely on Jackson instead of JSONB. Main reason is I have both REST (that uses Jackson) and then GraphQL (that uses JSONB) and thus ending up with two implementation for the same thing (so to say). Plus I found few differences in how they deal with different types for example dates - they expect different formats so that is a bit of an issue.

**Reply by t1** [reply]
What platform are you on? Can you switch to JSON-B? I think this would be the future-proof step to take, wouldn't it?

**Reply by mswiderski** [reply]
I am on quarkus. Firstly I was using JSON-B but changes pretty much at the same time when quarkus moved to Jackson as default one. And since it is used by some of the dependencies already it was just better to avoid duplicated libs for doing the same.

**Reply by phillip-kruger** [reply]
We will do some factory where one or the other can be loaded. In Quarkus we will load Jackson, in Wildfly Jsonb

**Reply by mswiderski** [reply]
excellent, looking forward to it!
